File: python/p052.py
'''
It can be seen that the number, 125874, and its double, 251748, contain exactly the same digits, but in a different order.

Find the smallest positive integer, x, such that 2x, 3x, 4x, 5x, and 6x, contain the same digits.

'''

from math import ceil, log10
from functools import reduce

def n_digits(n):
    return ceil(log10(n))

def isPerm(x, m):
    return reduce(lambda x, y: x == y, [sorted(str(x*k)) for k in range(1,m+1)])


for x in range(1, 6):
    for j in range(10**x, 1 + (10**(x+1) - 1)//6): # limits search based on number of digits matching
        if sorted(str(j)) == sorted(str(2*j)) == sorted(str(3*j)) == sorted(str(4*j)) == sorted(str(5*j)) == sorted(str(6*j)):
            print(j)
            break

File: cpp/p052.cpp
#include <iostream>
#include <cmath>
using namespace std;

int n_digits(long n);
bool is_perm(long n, long m);

int main() {
    for(int x = 1; x < 6; x++) {
        for(int j = int(pow(10, x)); j <= (pow(10,x+1) - 1)/6; j++) {
            if(is_perm(j, 6*j)) {
                bool found = true;
                for(int c = 2; c <= 6; c++) {
                    found = found && is_perm(j, c*j);
                }
                if(found) {
                    cout << j << endl;
                    return 1;
                }
            }
        }
    }
    return 0;
}

int n_digits(long n) {
    return 1 + floor(log10(n));
}

bool is_perm(long n, long m) {
    int d_n = n_digits(n);
    int d_m = n_digits(m);
    if(d_n != d_m)
        return false;
    else {
        int arr[10];
        for(int i = 0; i < 10; i++)
            arr[i] = 0;

        for(int j = 0; j < d_n; j++) {
            arr[ n % 10 ]++;
            arr[ m % 10 ]--;
            n /= 10;
            m /= 10;
        }
        for(int k = 0; k < 10; k++) {
            if(arr[k] != 0)
                return false;
        }
        return true;
    }
}
